### ISSUE TITLE
mc gives timeout problem

### ISSUE THREAD

**Issue by ridvankaratas** [minio/mc#2304]
I'm trying to mirror a huge bucket from s3 to my minio and it gives some errors several times for each folder in that bucket

```
mc: <ERROR> Failed to copy `https://s3.amazonaws.com/bucket-name/folder-name/tonya_p_ex_01/tonya_ex_0026.jpg`. Get https://bucket-name.s3.amazonaws.com/?delimiter=%2F&fetch-owner=true&list-type=2&max-keys=1000&prefix=folde-name%2Ftonya_p_ex_01%2Ftonya_ex_0026.jpg: dial tcp 52.216.20.235:443: i/o timeout
```

my mc version

```
Version: 2017-10-14T00:51:16Z
Release-tag: RELEASE.2017-10-14T00-51-16Z
Commit-id: 785e14a725357b39e22b74483cd202e7effa6195
```

System information
```
Centos7 - Linux  3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by ridvankaratas** [comment]
Also I got this when I try to `mc ls` on a folder inside a bucket which is located s3 

```dial tcp 54.231.40.82:443: i/o timeout```

**Comment by poornas** [comment]
@ridvankaratas, what is the size of the files being downloaded? this looks like a simple connection error since ```mc ls``` is also timing out.

**Comment by harshavardhana** [comment]
> dial tcp 54.231.40.82:443: i/o timeout

This is a common timeout behavior for AWS S3 on large buckets - we have 5 binomial retries but perhaps you might need to configure this to be a high value here. 

@poornas we can perhaps give a custom `mc` build with retries bumped to a high value??

**Comment by poornas** [comment]
@ridvankaratas , can you download this binary and try your uploads again (https://github.com/poornas/mc/blob/increaseretry/mc) - I increased the number of retries to 10.

**Comment by ridvankaratas** [comment]
@poornas its about 200GB and yes I will download that binary

**Comment by poornas** [comment]
@ridvankaratas , just hold on a sec - will upload the binary again

**Comment by poornas** [comment]
@ridvankaratas , pls use the binary on this link instead - https://github.com/poornas/mc/blob/master/mc

**Comment by ridvankaratas** [comment]
I downloaded and replaced binary with the old one. I got this when I try to run a mirror with the same buckets:
```
/usr/local/bin/mc: line 7: syntax error near unexpected token `newline'
/usr/local/bin/mc: line 7: `<!DOCTYPE html>'
```

**Comment by harshavardhana** [comment]
Looks like you downloaded HTML file. You might have to download again properly this time.

**Comment by ridvankaratas** [comment]
It gives this after I run mirror:

```
mc: <ERROR> Unable to stat source ... dial tcp 52.216.19.88:443: i/o timeout

mc: <ERROR> Unable to mirror objects ... dial tcp 52.216.232.27:443: i/o timeout
```

**Comment by ridvankaratas** [comment]
any commend or suggestion ?

**Comment by poornas** [comment]
@ridvankaratas, can you download the mc binary [here](https://github.com/poornas/mc/blob/20a4afdaa5936f3a850dc7e50db9042057778664/mc) and run your command again? I bumped up retries to 50 and added logging to help trouble shoot further.
please run  your mc mirror command with --debug flag and send me the output log.
```mc mirror --debug frombucket tobucket > log.txt``` 

Is your s3 region 'us-east-1' ?

**Comment by ridvankaratas** [comment]
```
mc: <DEBUG> Response Time:  322.900557ms

mc: <DEBUG> GET /mikandi-videos/?location= HTTP/1.1
Host: 127.0.0.1:9000
User-Agent: Minio (linux; amd64) minio-go/4.0.0 mc/2017-11-17T19:27:20Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20171118/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20171118T115626Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: application/xml
Date: Sat, 18 Nov 2017 11:56:26 GMT
Server: Minio/DEVELOPMENT.GOGET (linux; amd64)
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 14F82C1EDACFC46C

mc: <DEBUG> Response Time:  2.37253ms

mc: <DEBUG> GET /?fetch-owner=true&list-type=2&max-keys=1000&prefix=mojocontent%2FPureMissionary_HoloGirlsVR_Sample_GearVR_3dv.mp4%2F HTTP/1.1
Host: mikandi-videos.s3.amazonaws.com
User-Agent: Minio (linux; amd64) minio-go/4.0.0 mc/2017-11-17T19:27:20Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20171118/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20171118T115626Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Transfer-Encoding: chunked
Content-Type: application/xml
Date: Sat, 18 Nov 2017 11:56:27 GMT
Server: AmazonS3
X-Amz-Bucket-Region: us-east-1
X-Amz-Id-2: TZJM4iej7L8OutqxT1pUc5e0Y3eJwfDt6iK+qY8/ZZSBepDYhRf3w7gC04XzDiclVfCjU1kba58=
X-Amz-Request-Id: BDF421F2E9814CE3

mc: <DEBUG> Response Time:  77.857656ms

mc: <DEBUG> GET /mikandi-videos/?max-keys=1000&prefix=mojocontent%2FPureMissionary_HoloGirlsVR_Sample_GearVR_3dv.mp4%2F HTTP/1.1
Host: 127.0.0.1:9000
User-Agent: Minio (linux; amd64) minio-go/4.0.0 mc/2017-11-17T19:27:20Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20171118/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20171118T115626Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: application/xml
Date: Sat, 18 Nov 2017 11:56:27 GMT
Server: Minio/DEVELOPMENT.GOGET (linux; amd64)
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 14F82C1F1B795C8C

mc: <DEBUG> Response Time:  1.084454974s
```

**Comment by harshavardhana** [comment]
@ridvankaratas were you able to mitigate this problem?

**Comment by harshavardhana** [comment]
@ridvankaratas we added auto parallelization support in `mc` can you try latest master and see if it improves performance copy performance for you?

**Comment by ridvankaratas** [comment]
hi 
thanks for checking. it was still giving such error while doing mirror process but adding 50-retry helped with it. 

Okay I will copy the latest binary and test it

Appreciated!

**Comment by harshavardhana** [comment]
> hi
> thanks for checking. it was still giving such error while doing mirror process but adding 50-retry helped with it.
> 
> Okay I will copy the latest binary and test it
> 
> Appreciated!

So 50 retries work? @ridvankaratas  that is good to know. We can add an option to `mc` in such a scenario to make retries configurable. 

Please confirm if 50 retries binary worked fine.

**Comment by ridvankaratas** [comment]
yes it worked

**Comment by poornas** [comment]
@ridvankaratas, we are trying to tweak our retry count and backoff between retries so that it would work for  all users. I have added a MC_MAX_RETRIES env variable temporarily in this [mc binary](http://bit.ly/2BfvWuW). It would be very helpful if you could set this env variable and try your workload with the binary linked here.

What is the lowest number of retries that your mc cp/mirror operation succeeds? Earlier I had provided you a binary with max 10 retries(which didn't work for you) and most recently with 50. the env variable can be set with ```export MC_MAX_RETRIES=12``` if you are using linux.

**Comment by ridvankaratas** [comment]
@poornas thanks for this!

I will download that binary and use it. 

Actually 10 retries was working well, but it was still giving same error. We can say succeed rate as 75%

Honestly, I didn't track and examine retries. With this env variable, I guess I can track it. 

Thanks for the help!

**Comment by poornas** [comment]
@ridvankaratas , were you able to experiment with the retries? Appreciate your help on this.

**Comment by ridvankaratas** [comment]
hi

Yes I have. 

So this is what I got when I tried to mirror same bucket which I mirrored with the old binary before. I think its not a proper error? 

`mc: <ERROR> Unable to prepare URL for copying. Overwrite not allowed for`

**Comment by harshavardhana** [comment]
This is an okay error to ignore .. @ridvankaratas .this basically indicates that your source and destination files have changed but overwrite is not allowed by default. So you need to use --overwrite

**Comment by ridvankaratas** [comment]
ok I tried but now I have this:

`mc: <ERROR> Unable to prepare URL for copying. Delete not allowed for ...... `--remove` to override this behavior`

**Comment by harshavardhana** [comment]
Looks like you have files in destination too which are not present in source. So if needed you need to add `--remove` as well.

**Comment by poornas** [comment]
@ridvankaratas, you said earlier that with 10 retries, 75% of your mirror operations went through. Were you able to see what's the minimum number of retries that completes without timeout for your workload?

**Comment by harshavardhana** [comment]
> @ridvankaratas, you said earlier that with 10 retries, 75% of your mirror operations went through. Were you able to see what's the minimum number of retries that completes without timeout for your workload?

@poornas we can discuss with @abperiasamy about providing a higher retry limit? here. Should serve the purpose for other wider use cases.

**Comment by kannappanr** [comment]
Re-opening this issue as the fix in minio/minio-go#913  should be vendorized in mc for this issue to be marked closed.
